File: day2/typescript/main.ts
import { readFileSync } from "fs"


function isSafe(numbers: Array<number>) {
  let isCrescent = true;
  let isDecrescent = true;


  for (let i = 0; i < numbers.length - 1; i++) {
    let diff = Math.abs(numbers[i + 1] - numbers[i])

    if (diff > 3 || diff < 1) {
      return false
    }

    if (numbers[i + 1] < numbers[i]) {
      isCrescent = false
    }

    if (numbers[i + 1] > numbers[i]) {
      isDecrescent = false
    }
  }

  return isCrescent || isDecrescent
}


const solution = (path: string) => {
  let safeList = 0;
  try {
    const content = readFileSync(path, 'utf-8')
    const lines = content.split("\n");

    lines.forEach((line, index) => {
      const strNumbers = line.split(" ").filter((num) => num.trim() !== "");
      const numbers = strNumbers.map((num) => Number(num));

      if (numbers.length === 0) {
        return;
      }

      if (isSafe(numbers)) {
        safeList++
      } else {
        for (let i = 0; i < numbers.length; i++) {
          let temp = [...numbers.slice(0, i), ...numbers.slice(i + 1)]
          if (isSafe(temp)) {
            safeList++;
            break;
          }
        }
      }
    })
  } catch (err) {
    console.error("An error has ocurred", err)
  }

  console.log("Safe List:", safeList)
}


solution("./input.txt")

File: day2/lua/main.lua
local solution = {}
local safe_lists = 0

function Is_safe(numbers)
	local is_crescent = true
	local is_decrescent = true

	-- Iterate over numbers

	for i = 1, #numbers - 1 do
		local diff = math.abs(numbers[i + 1] - numbers[i])

		if diff > 3 or diff < 1 then
			return false
		end

		if numbers[i + 1] < numbers[i] then
			is_crescent = false
		end

		if numbers[i + 1] > numbers[i] then
			is_decrescent = false
		end
	end

	return is_crescent or is_decrescent
end

function solution:main(input)
	local file = io.open(input, "r")

	if not file then
		error("Could not open file " .. input)
	end

	for line in file:lines() do
		local numbers = {}

		for num in string.gmatch(line, "%d+") do
			table.insert(numbers, tonumber(num))
		end

		if Is_safe(numbers) then
			safe_lists = safe_lists + 1
		else
			for i = 1, #numbers do
				local temp = { table.unpack(numbers) }
				table.remove(temp, i)
				if Is_safe(temp) then
					safe_lists = safe_lists + 1
					break
				end
			end
		end
	end

	file:close()
	print("Safe Lists:", safe_lists)
end

solution:main("./input.txt")
